### ISSUE TITLE
Twisted should implement a reactor on top of asyncio

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8367]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8367|
|Type|enhancement|
|Created|2016-05-28 18:39:15Z|
|Branch|https://github.com/twisted/twisted/tree/aio-reactor-8367|

see: txtulip

This will allow Futures/Deferreds to interop a bit better, and maybe we can end up removing some reactors in 2020.

<details><summary>Searchable metadata</summary>

```
trac-id__8367 8367
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__aio_reactor_8367 aio-reactor-8367
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1464460755196831 1464460755196831
changetime__1470981420622670 1470981420622670
version__None None
owner__GitHub__noreply_____ GitHub <noreply@...>

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="graingert's avatar" src="https://avatars.githubusercontent.com/u/413772?s=50" width="50" height="50">](https://github.com/graingert)<a name="note_1"></a>|@graingert commented|
|-|-|

I've just packaged the txtulip repo, so that you can easily give this a go:

https://pypi.python.org/pypi/txtulip

It does seem to pass some of the current tests...

Maybe someone could try merging in the whole repo and seeing what happens?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Coverage seems to be mostly there as much as I can, putting it up for preliminary review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

PR at https://github.com/twisted/twisted/pull/446

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

See the test suite running with aio from start to finish: https://buildbot.twistedmatrix.com/boxes-unsupported?num_builds=2&&branch=aio-reactor-8367

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

Known issues: needs to gate it on py3 in twisted_reactors.py

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_7"></a>|@markrwilliams commented|
|-|-|

I dug into [the failing stdio test](https://github.com/twisted/twisted/pull/446/files#diff-8baf8e65e270dd6160eaebe5afe4f8cdR377).  Good news!  It's a bug in the standard library.

The `add_writer` method of asyncio event loops isn't atomic because all the Selector classes in the `selector` module internally record a file object before passing it to the platform's selector implementation.  If the platform's selector decides the file object isn't acceptable, the resulting exception doesn't cause the Selector to un-track the file object.

The failing/hanging stdio test goes through the following sequence of events (roughly):

1. The first `connection.write(intToByte(value))` call hits the asyncio reactor's `addWriter` method.
2. `addWriter` calls the asyncio loop's [add_writer method](https://github.com/python/cpython/blob/7098880f1fa9af6f9b85ac137ae4e77d01b3ded6/Lib/asyncio/selector_events.py#L267=L281), which happens to live on `_BaseSelectorEventLoop`.
3. The asyncio loop's `add_writer` method checks if the file object has been registered before via the selector's [get_key method](https://github.com/python/cpython/blob/7098880f1fa9af6f9b85ac137ae4e77d01b3ded6/Lib/asyncio/selector_events.py#L272).
4. It hasn't, so the `KeyError` block runs and [calls the selector's register method](https://github.com/python/cpython/blob/7098880f1fa9af6f9b85ac137ae4e77d01b3ded6/Lib/asyncio/selector_events.py#L236)
5. Code examples that follow use `EpollSelector`, but the code flow holds true for any other selector implementation. [
https://github.com/python/cpython/blob/7098880f1fa9af6f9b85ac137ae4e77d01b3ded6/Lib/selectors.py#L404-L412 The selector's register method] first calls through to the next `register` method in the MRO
6. That next method is always [_BaseSelectorImpl.register](https://github.com/python/cpython/blob/7098880f1fa9af6f9b85ac137ae4e77d01b3ded6/Lib/selectors.py#L233) which creates a `SelectorKey` instance for the file object, stores it under the file object's file descriptor, and then returns it.
7. Control returns to the concrete selector implementation, which [asks the operating system to track the file descriptor using the right API](https://github.com/python/cpython/blob/7098880f1fa9af6f9b85ac137ae4e77d01b3ded6/Lib/selectors.py#L411). 
8. The operating system refuses!  An exception is raised that, in this case, the asyncio reactor handles by creating a `_ContinuousPolling` object to watch the file descriptor.
9. The second `connection.write(intToByte(value))` call hits the asyncio reactor's `addWriter` method, which hits the `add_writer` method.  But the loop's selector's `get_key` method now returns a `SelectorKey`!  Now the asyncio reactor's `addWriter` method thinks the asyncio loop will watch the file descriptor, even though it won't.

This should be fixed upstream.  But one solution would be to unregister the file object with the selector after catching the exception.  That can be done with

```python
        except IOError as e:
            self._asyncioEventloop._selector.unregister(fd)

```

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_8"></a>|@markrwilliams commented|
|-|-|

THIS LOOKS REALLY GOOD!  I put another comment on Github.  Let me know if I can help with this!

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_9"></a>|@markrwilliams commented|
|-|-|

I say this is ready to merge.  The Buildbots disagree, though; I think their failures were spurious.  If there's a way to trigger a rebuild so that Github shows them as green, that'd be a good thing to do.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

That was my fault. reforced the builds.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|GitHub set owner to GitHub<br>GitHub set status to `closed`|
|-|-|

In changeset 1be0881a6fb5aab9464f87da82a6195bc46f5170
```
#!CommitTicketReference repository="" revision="1be0881a6fb5aab9464f87da82a6195bc46f5170"
Merge aio-reactor-8367: Add a asyncio reactor

Author: itamar, hawkowl
Reviewers: rodrigc, markrwilliams
Fixes: #8367
```
